plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+{- | Float bindings inwards. A binding can be floated if it is a lazy binding, or
+ a strict binding whose RHS is a value. Doing so does not have much impact on the
+ program size since it simply moves bindings around without changing the number
+ of bindings, but it may lead to more efficient evaluation.
+
+ Unlike the float-in in GHC, we /can/ float bindings into lambda abstractions,
+ because doing so cannot duplicate work. Therefore we simply float a binding
+ inwards to the deepest possible position.
+
+ We avoid moving a strict binding whose RHS is not a value, because doing so
+ may either change the semantics (if the RHS has effects) or duplicate work
+ (if it is floated into a lambda abstraction).
+
+ Floating inwards is done in a single pass. The subterms are processed bottom-up,
+ during which we keep track of the set of variable @Unique@s within each subterm.
+ The time complexity is linear wrt AST size (strictly speaking it is /O(n*log n)/
+ where the /log n/ comes from `Set` operations).
+
+ Currently we only move non-recursive term bindings. In the future we should also
+ implement float-in for other kinds of bindings.
+-}
+module PlutusIR.Transform.LetFloatIn (floatTerm) where
```

Comment:
no tests? :(

Reply:
I didn't know we have tests for individual optimizations. OK, I'll add some there.

Reply:
What about the edge cases I've mentioned in comments on this PR? There are lots of interesting cases specifically for float-in! And it's not that clear that the test for float-in are appropriate...

Reply:
... and the test cases for float-out include some comments that don't make any sense for float-in